server/utils/build/PushDockerTar.ts

```diff
@@ -0,0 +1,182 @@
+/* eslint-disable no-param-reassign */
+import { rm } from 'shelljs'
+import tar from 'tar'
+import { readFile, stat } from 'fs/promises'
+import fs from 'fs'
+import config from 'config'
+import { join } from 'path'
+import defaultAxios, { AxiosInstance } from 'axios'
+import { createHash } from 'crypto'
+import https from 'https'
+
+import { VersionDoc } from '../../models/Version'
+import { BuildOpts, BuildStep, Files } from './BuildStep'
+import { BuildLogger } from './BuildLogger'
+import { getAccessToken } from '../../routes/v1/registryAuth'
+import { ModelDoc } from '../../models/Model'
+
+const CHUNK_SIZE = 8 * 1024 * 1024
+
+export enum ContentTypes {
+  APPLICATION_OCTET_STREAM = 'application/octet-stream',
+  APPLICATION_MANIFEST = 'application/vnd.docker.distribution.manifest.v2+json',
+  APPLICATION_LAYER = 'application/vnd.docker.image.rootfs.diff.tar',
+  APPLICATION_CONFIG = 'application/vnd.docker.container.image.v1+json',
+}
+
+export type Layer = {
+  size: number
+  digest: string
+  mediaType: ContentTypes
+}
+
+export type Manifest = {
+  Config: string
+  RepoTags: string[]
+  Layers: string[]
+}
+
+async function getUploadUrl(axios: AxiosInstance, registry: string, image: string) {
+  const startUploadUrl = `${registry}/v2/${image}/blobs/uploads/`
+  const { location } = (await axios.post(startUploadUrl)).headers
+
+  return `${registry}${location}`
+}
+
+async function* readChunks(file: string, chunkSize: number): AsyncIterableIterator<Buffer> {
+  const readStream = fs.createReadStream(file, {
+    highWaterMark: chunkSize,
+  })
+  for await (const chunk of readStream) {
+    yield chunk
+  }
+}
+
+async function pushFile(axios: AxiosInstance, registry: string, uploadUrl: string, file: string) {
+  const sha256 = createHash('sha256')
+  let bytesRead = 0
+  let followUploadUrl = uploadUrl
+  let chunk
+  let headers
+  const { size } = await stat(file)
+
+  for await (chunk of readChunks(file, CHUNK_SIZE)) {
+    headers = {
+      'Content-Type': ContentTypes.APPLICATION_OCTET_STREAM,
+      'Content-Length': String(chunk.length),
+      'Content-Range': `${bytesRead}-${bytesRead + chunk.length}`,
+    }
+    bytesRead += chunk.length
+
+    sha256.update(chunk)
+
+    if (bytesRead < size) {
+      const { headers: responseHeaders } = await axios.patch(followUploadUrl, chunk, { headers })
+      followUploadUrl = `${registry}${responseHeaders.location}`
+    }
+  }
+
+  // last chunk
+  const digest = `sha256:${sha256.digest('hex')}`
+  await axios.put(`${followUploadUrl}&digest=${digest}`, chunk, { headers })
```

Comment:
Is there a reason we switch from using fetch to axios in different places? Is it worth being consist and using one, or does axios need to be used here?

Reply:
Yes, I agree we should aim to be consistent.  It seems that over time we've switched from `node-fetch` to `axios` so have defaulted to using that.  No technical reason beyond it having better support for progress, cancellation, etc.

Reply:
Happy to use axios in future work, I'll make a ticket for going through and making previous uses consistent.